src/transaction/flashback.c

```diff
@@ -0,0 +1,344 @@
+
+#include "flashback.h"
+
+#include "log_impl.h"
+#include "log_manager.h"
+#include "message_catalog.h"
+#include "msgcat_set_log.hpp"
+#include "memory_alloc.h"
+#include "intl_support.h"
+#include "log_record.hpp"
+#include "heap_file.h"
+#include "thread_entry.hpp"
+#include "oid.h"
+#include "storage_common.h"
+
+/*
+ * flashback_fill_dml_summary - fill the dml information into summary entry
+ *
+ * thread_p (in)             : thread entry
+ * summary_entry (in/out)    : flashback summary entry
+ * classoid (in)             : OID of DML target class
+ * rec_type(in)              : record type that can be classified as DML types
+ */
+
+static void
+flashback_fill_dml_summary (FLASHBACK_SUMMARY_ENTRY * summary_entry, OID classoid, SUPPLEMENT_REC_TYPE rec_type)
+{
+  assert (summary_entry != NULL);
+
+  bool found = false;
+  int i = 0;
+
+  /* fill in the summary_entry info entry
+   * 1. number of class and put the classoid into the classlist
+   * 2. number of DML operation */
+
+  if (summary_entry->classoid_set.find (classoid) == summary_entry->classoid_set.end ())
+    {
+      summary_entry->classoid_set.emplace (classoid);
+    }
+
+  switch (rec_type)
+    {
+    case LOG_SUPPLEMENT_INSERT:
+    case LOG_SUPPLEMENT_TRIGGER_INSERT:
+      summary_entry->num_insert++;
+      break;
+    case LOG_SUPPLEMENT_UPDATE:
+    case LOG_SUPPLEMENT_TRIGGER_UPDATE:
+      summary_entry->num_update++;
+      break;
+    case LOG_SUPPLEMENT_DELETE:
+    case LOG_SUPPLEMENT_TRIGGER_DELETE:
+      summary_entry->num_delete++;
+      break;
+    default:
+      assert (false);
+      break;
+    }
+
+  return;
+}
+
+/*
+ * flashback_make_summary - read log volumes within the range of start_lsa and end_lsa then make summary information
+ *
+ * return                : error code
+ * thread_p (in)         : thread entry
+ * context (in/out)      : flashback summary context that contains range of travering logs, filter (user, class), and summary list
+ */
+
+int
+flashback_make_summary_list (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_CONTEXT * context)
+{
+  LOG_PAGE *log_page_p = NULL;
+  char log_pgbuf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+
+  LOG_LSA cur_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA next_log_rec_lsa = LSA_INITIALIZER;
+  LOG_LSA process_lsa = LSA_INITIALIZER;
+
+  LOG_RECORD_HEADER *log_rec_header;
+
+  int trid;
+
+  time_t current_time = 0;
+
+  FLASHBACK_SUMMARY_ENTRY *summary_entry;
+  bool found = false;
+
+  int error = NO_ERROR;
+
+  LOG_RECTYPE log_type;
+
+  char *supplement_data = NULL;
+  int supplement_alloc_length = 0;
+
+  log_page_p = (LOG_PAGE *) PTR_ALIGN (log_pgbuf, MAX_ALIGNMENT);
+
+  assert (!LSA_ISNULL (&context->end_lsa));
+
+  LSA_COPY (&process_lsa, &(context->end_lsa));
+
+  /*fetch log page */
+  if (logpb_fetch_page (thread_p, &process_lsa, LOG_CS_SAFE_READER, log_page_p) != NO_ERROR)
+    {
+      /* er_set */
+      logpb_fatal_error (thread_p, false, ARG_FILE_LINE, "flashback_make_summary_list");
```

Comment:
why do you use logpb_fatal_error(), not er_set()?

Reply:
logpb_fetch_page() changes the return value to ER_FAILED and sends it without er_set (), regardless of any internal error. 
Since I do not know what kind of error occurred internally, I used logpb_fatal_error(), 
which is also used in vacuum and recovery.

how about 
```
error = logpb_fetch_page ()
if (error != NO_ERROR) 
{ logpb_fatal_error ()..}
```